### PR TITLE
[SPARK-31013][Core][WebUI] InMemoryStore: improve removeAllByIndexValues over natural key index

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

The method `removeAllByIndexValues` in KVStore is to delete all the objects which have certain values in the given index.
However, in the current implementation of `InMemoryStore`, when the given index is the natural key index, there is no special handling for it and a linear search over all the task data is performed.

We can improve it by deleting the natural keys directly from the internal hashmap.

### Why are the changes needed?

Better performance if the given index for `removeAllByIndexValues` is the natural key index in 
`InMemoryStore`
### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Enhance the existing test.

### CODE REVIEW

**Comment by gengliangwang** [comment]
retest this please.

File: common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java [hunk 1]
```diff
@@ -271,9 +276,9 @@ public void put(T value) throws Exception {
       }
     }
 
-    public void delete(Object key) {
-      data.remove(asKey(key));
-      if (hasNaturalParentIndex) {
+    public Boolean delete(Object key) {
```
Comment (HeartSaVioR) [inline common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java]: At a glance, it could be primitive `boolean` here and below.

Comment (gengliangwang) [reply common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java]: Ah, right.

**Review by HeartSaVioR — APPROVED** [review]
LGTM.

**Comment by HeartSaVioR** [comment]
Btw, out-of-topic, but while we are here, is it intentional or missing spot CountingRemoveIfForEach only removes the data and doesn't touch parentToChildrenMap? If it's just a missing spot maybe I could submit a PR to fix it.

**Comment by gengliangwang** [comment]
> Btw, out-of-topic, but while we are here, is it intentional or missing spot CountingRemoveIfForEach only removes the data and doesn't touch parentToChildrenMap? If it's just a missing spot maybe I could submit a PR to fix it.


You are right. It is a missing part of https://github.com/apache/spark/pull/27716

**Comment by HeartSaVioR** [comment]
Thanks for confirming, @gengliangwang !

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
Just submitted a PR #27765. It might be possible to be conflicted - I'll rebase if necessary once this PR gets merged.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**